File: 4code.py
import random

# Dictionary to store employee shift preferences
employee_preferences = {
    "Monday": [("John", "morning"), ("Alice", "afternoon"), ("Bob", "evening")],
    "Tuesday": [("John", "evening"), ("Alice", "morning"), ("Bob", "afternoon")],
    "Wednesday": [("John", "morning"), ("Alice", "afternoon"), ("Bob", "evening")],
    # Add for all days of the week...
}

# Final schedule with shifts per day
final_schedule = {
    "Monday": {"morning": [], "afternoon": [], "evening": []},
    "Tuesday": {"morning": [], "afternoon": [], "evening": []},
    "Wednesday": {"morning": [], "afternoon": [], "evening": []},
    "Thursday": {"morning": [], "afternoon": [], "evening": []},
    "Friday": {"morning": [], "afternoon": [], "evening": []},
    "Saturday": {"morning": [], "afternoon": [], "evening": []},
    "Sunday": {"morning": [], "afternoon": [], "evening": []},
}

# Track number of shifts each employee has been assigned to
employee_shifts = {}

def assign_shift(day, shift, employee):
    """Assign an employee to a shift if they haven't worked 5 days yet."""
    if employee_shifts.get(employee, 0) < 5:
        final_schedule[day][shift].append(employee)
        employee_shifts[employee] = employee_shifts.get(employee, 0) + 1
        return True
    return False

def resolve_conflict(day, shift, employee):
    """Resolve shift conflicts by reassigning to another available shift."""
    if len(final_schedule[day][shift]) < 2:
        # Try to find an alternative shift for the employee
        for alt_shift in ["morning", "afternoon", "evening"]:
            if alt_shift != shift and len(final_schedule[day][alt_shift]) < 2:
                if assign_shift(day, alt_shift, employee):
                    print(f"{employee} assigned to {alt_shift} on {day} (original shift conflict)")
                    return True
    return False

# Example of assigning employees based on their preferences
for day, preferences in employee_preferences.items():
    for employee, preferred_shift in preferences:
        if not assign_shift(day, preferred_shift, employee):
            if not resolve_conflict(day, preferred_shift, employee):
                print(f"Could not assign {employee} to any shift on {day}")

# Output the final schedule
def print_schedule():
    for day, shifts in final_schedule.items():
        print(f"\n{day}:")
        for shift, employees in shifts.items():
            print(f"  {shift.capitalize()}: {', '.join(employees)}")

print_schedule()

File: 4code.java
import java.util.*;

public class EmployeeScheduler {

    static Map<String, Map<String, List<String>>> finalSchedule = new HashMap<>();
    static Map<String, Integer> employeeShifts = new HashMap<>();

    public static void main(String[] args) {
        // Initialize schedule structure
        String[] days = {"Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"};
        String[] shifts = {"morning", "afternoon", "evening"};
        
        for (String day : days) {
            finalSchedule.put(day, new HashMap<>());
            for (String shift : shifts) {
                finalSchedule.get(day).put(shift, new ArrayList<>());
            }
        }

        // Example employee preferences (could be input dynamically)
        List<Map<String, String>> employeePreferences = new ArrayList<>();
        employeePreferences.add(Map.of("day", "Monday", "name", "John", "shift", "morning"));
        employeePreferences.add(Map.of("day", "Monday", "name", "Alice", "shift", "afternoon"));
        // Add more preferences...

        // Process assignments
        for (Map<String, String> pref : employeePreferences) {
            String day = pref.get("day");
            String employee = pref.get("name");
            String preferredShift = pref.get("shift");

            if (!assignShift(day, preferredShift, employee)) {
                resolveConflict(day, preferredShift, employee);
            }
        }

        // Print the final schedule
        printSchedule();
    }

    public static boolean assignShift(String day, String shift, String employee) {
        if (employeeShifts.getOrDefault(employee, 0) < 5) {
            finalSchedule.get(day).get(shift).add(employee);
            employeeShifts.put(employee, employeeShifts.getOrDefault(employee, 0) + 1);
            return true;
        }
        return false;
    }

    public static void resolveConflict(String day, String shift, String employee) {
        if (finalSchedule.get(day).get(shift).size() < 2) {
            for (String altShift : finalSchedule.get(day).keySet()) {
                if (!altShift.equals(shift) && finalSchedule.get(day).get(altShift).size() < 2) {
                    if (assignShift(day, altShift, employee)) {
                        System.out.println(employee + " assigned to " + altShift + " on " + day + " (original shift conflict)");
                        return;
                    }
                }
            }
        }
    }

    public static void printSchedule() {
        for (String day : finalSchedule.keySet()) {
            System.out.println("\n" + day + ":");
            for (String shift : finalSchedule.get(day).keySet()) {
                System.out.println("  " + shift.substring(0, 1).toUpperCase() + shift.substring(1) + ": " + String.join(", ", finalSchedule.get(day).get(shift)));
            }
        }
    }
}
